### ISSUE TITLE
[Feature Request] Support `@beartype` + `@contextlib.contextmanager` decoration chaining

### ISSUE THREAD

**Issue by leycec** [beartype/beartype#249]
En-route to the pit of doom known as `beartype.claw`, @qutebrowser maestro @The-Compiler uncovered a [horrible ordering issue when chaining the `@beartype` decorator with the standard `@contextlib.contextmanager` decorator](https://github.com/beartype/beartype/issues/43#issuecomment-1630752667).

## Take It Away, Maestro!

Welp, I believe I ran into another interesting corner case - here is a minimal example:

```python
from collections.abc import Iterator
from contextlib import contextmanager
from beartype import beartype

@beartype  # this breaks, and is what beartype.claw does
@contextmanager
# @beartype  # this works
def ctx() -> Iterator[None]:
    yield

with ctx():
    pass
```

which results in (newlines inserted for readability):

```python
Traceback (most recent call last):
  File ".../test.py", line 11, in <module>
    with ctx():
  File "<@beartype(__main__.ctx) at 0x7f6e76793be0>", line 19, in ctx
beartype.roar.BeartypeCallHintReturnViolation: Function __main__.ctx()
   return <contextlib._GeneratorContextManager object at 0x7f6e767a1570>
   violates type hint collections.abc.Iterator[None],
   as <protocol ABC "contextlib._GeneratorContextManager"> <contextlib._GeneratorContextManager object at 0x7f6e767a1570>
   not instance of <protocol ABC "collections.abc.Iterator">.
```

Which is true when looking at the decorated function. However, we're annotating the undecorated one. Moving the `@beartype` to after (lines-wise) before (time-wise, as Python runs decorators from innermost to outermost) `@contextmanager` fixes this, but obviously that isn't possible with `beartype.claw`.

## What @leyceyc Has To Say: *Nuthin' Much*

@beartype will *probably* need to accommodate this like it does similar ordering issues in decorating chaining with the standard `@classmethod`, `@property`, and `@staticmethod` decorators – namely, by forcefully reversing the order of decoration. 

Thankfully, this means that this is actually an issue with the `@beartype` decorator rather than the `beartype.claw` API. Ideally, `@beartype` should:

1. Detect when `@beartype` is chained *after* `@contextlib.contextmanager`.
2. When this is detected:
   1. Get the original callable decorated by `@contextlib.contextmanager`.
   2. Decorate that (rather than the callable created by `@contextlib.contextmanager`) by `@beartype`.
   3. Re-decorate the callable created by `@beartype` by `@contextlib.contextmanager`, effectively "throwing away" the prior callable created by `@contextlib.contextmanager`.

This implies a bit of decoration-time inefficiency. But... we currently do that with `@classmethod`, `@property`, and `@staticmethod` decorators and nobody particularly minds. Except they *will*, now that I've publicly admitted to our crimes. :sweat_smile:

**Comment by The-Compiler** [comment]
Note that I don't think this is specific to `@contextlib.contextmanager`. For example, with this beauty:

```python
import functools
from collections.abc import Callable
from beartype import beartype

def actually_return_the_length_oh_how_useful(func: Callable[[], str]) -> Callable[[], int]:
    @functools.wraps(func)
    def wrapper() -> int:
        return len(func())
    return wrapper

@beartype
@actually_return_the_length_oh_how_useful
def func() -> str:
    return "roar"

func()
```

beartype also complains that:

```python
Traceback (most recent call last):
  File ".../test.py", line 16, in <module>
    func()
  File "<@beartype(__main__.func) at 0x7f328f8320e0>", line 18, in func
beartype.roar.BeartypeCallHintReturnViolation: Function __main__.func() return 4 violates type hint <class 'str'>, as int 4 not instance of str.
```

but in reality, the code is correct, Python is happy, and so is mypy (other than, ironically, `@beartype` not being typed and thus making the function untyped :sweat_smile:).

**Comment by leycec** [comment]
*...heh.* I acknowledge your brutal toy example publicly highlighting our badness. Since mypy silently accepts that, so should we. Mypy isn't wrong, either; we are.

For simplicity, @beartype currently repeatedly unwraps callables decorated by `@functools.wraps` until "bottoming out" (i.e., eventually finding a callable *not* decorated by `@functools.wraps`). At the time, that seemed reasonable – and mostly was. But times change. Sadly, so do we. I, for example, am now bald.

@beartype should likewise change. Specifically, @beartype should follow a more nuanced heuristic for detecting the decorated callable of interest:

1. If the decorated callable is both **wrapped** (i.e., decorated by the `@functools.wraps` decorator) *and* accepts only the standard variadic arguments `*args` and `**kwargs`, then the decorated callable is almost certainly an **isomorphic decorator closure** (i.e., closure created and returned by a standard decorator preserving the position and types of all passed parameters and returns). In this case, @beartype should:
   1. Unwrap this closure once.
   2. Repeat the outermost **step 1.** above.
2. Else, the decorated callable is a non-isomorphic decorator closure of some sort that is *probably* transforming the position and types of one or more passed parameters or returns. In this case, @beartype should just decorate this closure as is *without* unwrapping this closure at all.

Well, that intensified rapidly.

**Comment by leycec** [comment]
...lolbro. I implemented all this extraneous functionality with your example as a test case – only to belatedly realize that both your example *and* mypy are painfully busted. The odds were against us, but @beartype is the only thing that is telling the truth here. I am equally shocked.

Allow me to pontificate in a tedious manner. This is the problem in your example:

```python
    @functools.wraps(func)
    def wrapper() -> int:
```

It's hard to see the problem until you've seen the problem. You've annotated a function with the return type hint `int` but then decorated that function by  the `@functools.wraps()` decorator, which then silently propagates (i.e., replaces) *all* type hints annotating the wrappee function `func` onto the `wrapper`. Specifically, `@functools.wraps()` replaces the return type hint `int` annotating `wrapper` with the return type hint `str` annotating `func`.

That's the whole *raison d'être* of `@functools.wraps()`, after all: to propagate metadata from the wrappee onto the wrapper. `@functools.wraps()` should at least emit a warning when the wrapper is already annotated by one or more type hints, because silently destroying type hints is badness. Instead, `@functools.wraps()` silently destroys type hints. I sigh.

All of this hand-waving means that your example is semantically equivalent to:

```python
    @functools.wraps(func)
    def wrapper() -> str:
```

Right? The `int` return type hint has been destroyed by `@functools.wraps()` and replaced by the original `str` return type hint, which @beartype then correctly sees and complains about.

In short, **PEBKAC.** Also, mypy is wrong. Who didn't see *that* one coming? `</sigh>`

**Comment by The-Compiler** [comment]
> Specifically, `@functools.wraps()` replaces the return type hint `int` annotating `wrapper` with the return type hint `str` annotating `func`.

A convincing argument! As you say indeed,

```python
import typing
import functools

def ret_int() -> int:
    pass

@functools.wraps(ret_int)
def ret_str() -> str:
    pass

print(typing.get_type_hints(ret_str))
```

prints

```python
{'return': <class 'int'>}
```

It looks like this can get overridden via a rather funny-looking:

```python
@functools.wraps(ret_int, assigned=set(functools.WRAPPER_ASSIGNMENTS) - {'__annotations__'})
```

but I've yet to see someone who actually does that!

---

As for:

> If an IDE is complaining about these things, that IDE is dumb and best ignored. IDEs are only as smart as static type-checkers and static type-checkers are dumb. This is why my face hurts.

it was mypy complaining about it. I was ready to blame beartype for a missing `py.typed` marker, but I see that it's actually there, so admittedly I have no idea what's going on.

**Comment by leycec** [comment]
**Resolved by 2b8c6823742...** *sorta.* Under:

* Python ≥ 3.11, @beartype now supports this seemingly perfectly. Would tests lie? <sup>\*ducks\*</sup>
* Python ≤ 3.10, @beartype doesn't support this at all. It's a long and tiresome story. My face is already falling asleep just thinking about it.

Basically, the closure returned by `@contextmanager` masquerades as the original generator very well – a little *too* well, in fact. Under Python ≥ 3.11, we can break that masquerade by inspecting the `co_qualname` attribute of that closure. Under Python ≤ 3.10, `co_qualname` doesn't exist and there's no reasonable means of fabricating it; our only recourse is to (*...waitforit*) politely request that users explicitly decorate `@beartype` *before* `@contextmanager` under Python ≤ 3.10.

Thankfully, the `beartype.claw` API detects when a class or callable has already been decorated by `@beartype` and avoids re-doing so. It's still a pain to manually decorate *every* context manager by `@beartype`, but at least it's technically feasible.

## Here Is Where I Thank You

Thanks a heap for double-checking my madness, @The-Compiler. I'm usually certifiably wrong about everything, including the answer to that most troubling of questions: "Does this dress make me look very fat *or* only partially fat?"

I'm unsurprised that the @qutebrowser mastermind also knew about this hidden darkness...

> It looks like this can get overridden via a rather funny-looking:

```python
@functools.wraps(ret_int, assigned=set(functools.WRAPPER_ASSIGNMENTS) - {'__annotations__'})
```

> but I've yet to see someone who actually does that!

...you win this feature request. That's *extremely* clever. In theory, @beartype 0.14.1 should already support those kinds of unfathomable shenanigans out-of-the-box. In practice, I have tested nothing and promise even less.

We now return to everyone's regularly scheduled weekend coding marathon. :smiling_face_with_tear:

**Comment by gabrieldemarmiesse** [comment]
I add here for completeness that if you want typebear to only be a test-time dependency you can use the `@typing.no_type_check` decorator when you are using python < 3.11. It will disable the check for this context manager.
